packages/actor-extract-links-extract-tree/lib/SolutionDomain.ts

```diff
@@ -0,0 +1,198 @@
+import { SolutionRange } from './SolutionRange';
+import { LogicOperator } from './solverInterfaces';
+
+const AN_EMPTY_SOLUTION_RANGE = new SolutionRange(undefined);
+/**
+ * A class representing the domain of a solution of system of boolean equation.
+ * Every operation return a new object.
+ */
+export class SolutionDomain {
+  /**
+     * The multiple segment of the domain, it is always order by the lower bound
+     * of the SolutionRange.
+     */
+  private domain: SolutionRange[] = [];
+
+  /**
+    * The last operation apply to the domain
+    *
+    */
+  private lastOperation: LogicOperator | undefined = undefined;
+
+  /**
+     * Get the multiple segment of the domain.
+     * @returns {SolutionRange[]}
+     */
+  public get_domain(): SolutionRange[] {
+    return new Array(...this.domain);
```

Comment:
Why are you copying the array?

Reply:
To avoid problems with shallow copying.

Reply:
Where do these problems occur?

Reply:
I want to have an immutable `SolutionDomain`, so if I have a shallow copy it is actually mutable, it doesn't happen with my code but with the context object I already had this problem because I forgot that javascript was doing shallow copies even though it was supposedly an immutable object.

Reply:
But where do you mutate it? Perhaps the problem lies in that part of the code then?

Reply:
I don't mutate it, I want to be sure that it cannot be mutated.